examples/FisherPrice_cu/cufisher_price_v2.cu

```diff
@@ -0,0 +1,231 @@
+  static Scoring *MakeInstanceAt(void *addr)
+  {
+    Scoring *obj = new (addr) Scoring();
+    return obj;
+  }
+};
+
+// kernel function that does transportation
+__global__ void transport(int n, adept::BlockData<MyTrack> *block, curandState_t *states, Queue_t *queues)
+{
+  for (int i = blockIdx.x * blockDim.x + threadIdx.x; i < n; i += blockDim.x * gridDim.x) {
+    // transport particles
+    for (int xyz = 0; xyz < 3; xyz++) {
+      (*block)[i].pos[xyz] = (*block)[i].pos[xyz] + (*block)[i].energy * (*block)[i].dir[xyz];
+    }
+  }
+}
+
+// kernel function that assigns next process to the particle
+__global__ void select_process(adept::BlockData<MyTrack> *block, Scoring *scor, curandState_t *states,
+                               Queue_t *queues[])
+{
+  int particle_index = blockIdx.x * blockDim.x + threadIdx.x;
+
+  // check if you are not outside the used block
+  if (particle_index > block->GetNused() + block->GetNholes()) return;
+
+  // check if the particle is still alive (E>0)
+  if ((*block)[particle_index].energy == 0) return;
+
+  // generate random number
+  float r = curand_uniform(states);
+
+  if (r > 0.5f) {
+    queues[0]->enqueue(particle_index);
+  } else {
+    queues[1]->enqueue(particle_index);
+  }
+}
+
+// kernel function that does energy loss
+__global__ void process_eloss(int n, adept::BlockData<MyTrack> *block, Scoring *scor, curandState_t *states,
+                              Queue_t *queue)
+{
+  int particle_index;
+  for (int i = blockIdx.x * blockDim.x + threadIdx.x; i < n; i += blockDim.x * gridDim.x) {
+    if (!queue->dequeue(particle_index)) return;
+
+    // check if the particle is still alive (E>0)
+    if ((*block)[particle_index].energy == 0) return;
+
+    // call the 'process'
+    // energy loss
+    float eloss = 0.2f * (*block)[particle_index].energy;
+    scor->totalEnergyLoss.fetch_add(eloss < 0.001f ? (*block)[particle_index].energy : eloss);
+    (*block)[particle_index].energy = (eloss < 0.001f ? 0.0f : ((*block)[particle_index].energy - eloss));
+
+    // if particle dies (E=0) release the slot
+    if ((*block)[particle_index].energy < 0.001f) block->ReleaseElement(particle_index);
+  }
+}
+
+// kernel function that does pair production
+__global__ void process_pairprod(int n, adept::BlockData<MyTrack> *block, Scoring *scor, curandState_t *states,
+                                 Queue_t *queue)
+{
+  int particle_index;
+  for (int i = blockIdx.x * blockDim.x + threadIdx.x; i < n; i += blockDim.x * gridDim.x) {
+    if (!queue->dequeue(particle_index)) return;
+
+    // check if the particle is still alive (E>0)
+    if ((*block)[particle_index].energy == 0) return;
+
+    // pair production
+    float eloss = 0.5f * (*block)[particle_index].energy;
+    (*block)[particle_index].energy -= eloss;
+
+    // here I need to create a new particle
+    auto secondary_track = block->NextElement();
+    assert(secondary_track != nullptr && "No slot available for secondary track");
```

Comment:
Should we have the creation attempt+assertion prior to the calculations? That way, if we replace the assertion with a later "couldn't process this" like error/return, we don't change any physical state of the particle.

Reply:
Yes indeed. I have changed the order to be more consistent.